### ISSUE TITLE
systemd-sysctl does not disable IPv6 on physical network devices during boot

### ISSUE THREAD

**Issue by derolidev** [systemd/systemd#33762]
### systemd version the issue has been seen with

255

### Used distribution

Ubuntu 24.04 LTS

### Linux kernel version used

6.8.0-38-generic

### CPU architectures issue was seen on

None

### Expected behaviour you didn't see

Creating a configuration file `/etc/sysctl.d/1000-disable-ipv6.conf` should disable IPv6 on all network devices.

```bash
setup@server1:~$ sudo cat /etc/sysctl.d/1000-disable-ipv6.conf 
net.ipv6.conf.all.disable_ipv6 = 1
net.ipv6.conf.default.disable_ipv6 = 1
net.ipv6.conf.lo.disable_ipv6 = 1
```

```bash
setup@server1:~$ cat /proc/sys/net/ipv6/conf/all/disable_ipv6
1
```

```bash
setup@server1:~$ sudo grep -r . /proc/sys/net/ipv6/conf/*/disable_ipv6
/proc/sys/net/ipv6/conf/all/disable_ipv6:1
/proc/sys/net/ipv6/conf/default/disable_ipv6:1
/proc/sys/net/ipv6/conf/ens33/disable_ipv6:1
/proc/sys/net/ipv6/conf/lo/disable_ipv6:1
```

It seems that `systemd-sysctl` is started before physical network interfaces are present.

### Unexpected behaviour you saw

IPv6 is not disabled on physical network devices, like `ens33`.

```bash
setup@server1:~$ sudo cat /proc/sys/net/ipv6/conf/all/disable_ipv6
1
```

```bash
setup@server1:~$ sudo grep -r . /proc/sys/net/ipv6/conf/*/disable_ipv6
/proc/sys/net/ipv6/conf/all/disable_ipv6:1
/proc/sys/net/ipv6/conf/default/disable_ipv6:1
/proc/sys/net/ipv6/conf/ens33/disable_ipv6:0
/proc/sys/net/ipv6/conf/lo/disable_ipv6:1
```

### Steps to reproduce the problem

1. Create a configuration `/etc/sysctl.d/1000-disable-ipv6.conf` with content below.
```bash
net.ipv6.conf.all.disable_ipv6 = 1
net.ipv6.conf.default.disable_ipv6 = 1
net.ipv6.conf.lo.disable_ipv6 = 1
```
2. Reboot the server.
3. Run `cat /proc/sys/net/ipv6/conf/all/disable_ipv6`. It will output `1`.
4. Run `sudo grep -r . /proc/sys/net/ipv6/conf/*/disable_ipv6`. It will return the result below. Notice `/proc/sys/net/ipv6/conf/ens33/disable_ipv6:0`.
```bash
/proc/sys/net/ipv6/conf/all/disable_ipv6:1
/proc/sys/net/ipv6/conf/default/disable_ipv6:1
/proc/sys/net/ipv6/conf/ens33/disable_ipv6:0
/proc/sys/net/ipv6/conf/lo/disable_ipv6:1
```
5. Run `ip a`. It will return the result below. Notice `inet6 fe80::20c:29ff:fe58:7a05/64 scope link ` on `ens33`.
```
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
2: ens33: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP group default qlen 1000
    link/ether 00:0c:29:58:7a:05 brd ff:ff:ff:ff:ff:ff
    altname enp2s1
    inet 10.177.114.104/24 metric 100 brd 10.177.114.255 scope global dynamic ens33
       valid_lft 2906sec preferred_lft 2906sec
    inet6 fe80::20c:29ff:fe58:7a05/64 scope link 
       valid_lft forever preferred_lft forever
```
6.  Run `sudo sysctl -p /etc/sysctl.d/1000-disable-ipv6.conf` or `sudo sysctl --system` manually.
7. Run `sudo grep -r . /proc/sys/net/ipv6/conf/*/disable_ipv6`. It will return the result below. Notice `/proc/sys/net/ipv6/conf/ens33/disable_ipv6:1`.
```bash
/proc/sys/net/ipv6/conf/all/disable_ipv6:1
/proc/sys/net/ipv6/conf/default/disable_ipv6:1
/proc/sys/net/ipv6/conf/ens33/disable_ipv6:1
/proc/sys/net/ipv6/conf/lo/disable_ip
```
8. Run `ip a`. It will return the result below. Notice `inet6 fe80::20c:29ff:fe58:7a05/64 scope link ` is gone on `ens33`.
```
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
2: ens33: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP group default qlen 1000
    link/ether 00:0c:29:58:7a:05 brd ff:ff:ff:ff:ff:ff
    altname enp2s1
    inet 10.177.114.104/24 metric 100 brd 10.177.114.255 scope global dynamic ens33
       valid_lft 2906sec preferred_lft 2906sec
```

**Comment by poettering** [comment]
what do you run on "ens33" to manage the interface? networkd? if so, it will manage the interface comprehensively, including sysctls. Thus you are just conflciting with that.

**Comment by derolidev** [comment]
The server is running with the default-packages, which are shipped with Ubuntu 24.04 LTS Server.
Configuration of the network is done via `neplan` which seems to use `systemd-networkd`.

```bash
setup@server1:~$ sudo systemctl status systemd-networkd.service
[sudo] password for setup:
● systemd-networkd.service - Network Configuration
     Loaded: loaded (/usr/lib/systemd/system/systemd-networkd.service; enabled; preset: enabled)
     Active: active (running) since Sun 2024-07-21 12:59:51 UTC; 21s ago
TriggeredBy: ● systemd-networkd.socket
       Docs: man:systemd-networkd.service(8)
             man:org.freedesktop.network1(5)
   Main PID: 637 (systemd-network)
     Status: "Processing requests..."
      Tasks: 1 (limit: 4556)
   FD Store: 0 (limit: 512)
     Memory: 3.7M (peak: 4.0M)
        CPU: 44ms
     CGroup: /system.slice/systemd-networkd.service
             └─637 /usr/lib/systemd/systemd-networkd

Jul 21 12:59:51 server1 systemd[1]: Starting systemd-networkd.service - Network Configuration...
Jul 21 12:59:51 server1 systemd-networkd[637]: lo: Link UP
Jul 21 12:59:51 server1 systemd-networkd[637]: lo: Gained carrier
Jul 21 12:59:51 server1 systemd-networkd[637]: Enumeration completed
Jul 21 12:59:51 server1 systemd[1]: Started systemd-networkd.service - Network Configuration.
Jul 21 12:59:51 server1 systemd-networkd[637]: ens33: Configuring with /run/systemd/network/10-netplan-ens33.network.
Jul 21 12:59:51 server1 systemd-networkd[637]: ens33: Link UP
Jul 21 12:59:51 server1 systemd-networkd[637]: ens33: Gained carrier
Jul 21 12:59:51 server1 systemd-networkd[637]: ens33: DHCPv4 address 10.177.114.104/24, gateway 10.177.114.12 acquired from 10.177.114.13
Jul 21 12:59:53 server1 systemd-networkd[637]: ens33: Gained IPv6LL
```

```bash
setup@server1:~$ sudo systemctl --type=service
  UNIT                                                                                      LOAD   ACTIVE SUB     DESCRIPTION
  apparmor.service                                                                          loaded active exited  Load AppArmor profiles
  apport.service                                                                            loaded active exited  automatic crash report generation
  blk-availability.service                                                                  loaded active exited  Availability of block devices
  console-setup.service                                                                     loaded active exited  Set console font and keymap
  cron.service                                                                              loaded active running Regular background program processing daemon
  dbus.service                                                                              loaded active running D-Bus System Message Bus
  finalrd.service                                                                           loaded active exited  Create final runtime dir for shutdown pivot root
  getty@tty1.service                                                                        loaded active running Getty on tty1
  keyboard-setup.service                                                                    loaded active exited  Set the console keyboard layout
  kmod-static-nodes.service                                                                 loaded active exited  Create List of Static Device Nodes
  lvm2-monitor.service                                                                      loaded active exited  Monitoring of LVM2 mirrors, snapshots etc. using dmeventd or progress polling
  ModemManager.service                                                                      loaded active running Modem Manager
  multipathd.service                                                                        loaded active running Device-Mapper Multipath Device Controller
  open-vm-tools.service                                                                     loaded active running Service for virtual machines hosted on VMware
  plymouth-quit-wait.service                                                                loaded active exited  Hold until boot process finishes up
  plymouth-quit.service                                                                     loaded active exited  Terminate Plymouth Boot Screen
  plymouth-read-write.service                                                               loaded active exited  Tell Plymouth To Write Out Runtime Data
  polkit.service                                                                            loaded active running Authorization Manager
  rsyslog.service                                                                           loaded active running System Logging Service
  setvtrgb.service                                                                          loaded active exited  Set console scheme
  snapd.apparmor.service                                                                    loaded active exited  Load AppArmor profiles managed internally by snapd
  snapd.seeded.service                                                                      loaded active exited  Wait until snapd is fully seeded
  ssh.service                                                                               loaded active running OpenBSD Secure Shell server
  sysstat.service                                                                           loaded active exited  Resets System Activity Logs
  systemd-binfmt.service                                                                    loaded active exited  Set Up Additional Binary Formats
  systemd-fsck@dev-disk-by\x2duuid-ed758a72\x2d9acb\x2d4ff8\x2da152\x2d7584c2144bf4.service loaded active exited  File System Check on /dev/disk/by-uuid/ed758a72-9acb-4ff8-a152-7584c2144bf4
  systemd-journal-flush.service                                                             loaded active exited  Flush Journal to Persistent Storage
  systemd-journald.service                                                                  loaded active running Journal Service
  systemd-logind.service                                                                    loaded active running User Login Management
  systemd-modules-load.service                                                              loaded active exited  Load Kernel Modules
  systemd-networkd-wait-online.service                                                      loaded active exited  Wait for Network to be Configured
  systemd-networkd.service                                                                  loaded active running Network Configuration
  systemd-random-seed.service                                                               loaded active exited  Load/Save OS Random Seed
  systemd-remount-fs.service                                                                loaded active exited  Remount Root and Kernel File Systems
  systemd-resolved.service                                                                  loaded active running Network Name Resolution
  systemd-sysctl.service                                                                    loaded active exited  Apply Kernel Variables
  systemd-timesyncd.service                                                                 loaded active running Network Time Synchronization
  systemd-tmpfiles-setup-dev-early.service                                                  loaded active exited  Create Static Device Nodes in /dev gracefully
  systemd-tmpfiles-setup-dev.service                                                        loaded active exited  Create Static Device Nodes in /dev
  systemd-tmpfiles-setup.service                                                            loaded active exited  Create Volatile Files and Directories
  systemd-udev-trigger.service                                                              loaded active exited  Coldplug All udev Devices
  systemd-udevd.service                                                                     loaded active running Rule-based Manager for Device Events and Files
  systemd-update-utmp.service                                                               loaded active exited  Record System Boot/Shutdown in UTMP
  systemd-user-sessions.service                                                             loaded active exited  Permit User Sessions
  udisks2.service                                                                           loaded active running Disk Manager
  ufw.service                                                                               loaded active exited  Uncomplicated firewall
  unattended-upgrades.service                                                               loaded active running Unattended Upgrades Shutdown
  user-runtime-dir@1000.service                                                             loaded active exited  User Runtime Directory /run/user/1000
  user@1000.service                                                                         loaded active running User Manager for UID 1000
```

**Comment by derolidev** [comment]
We can close this one then. When `netplan` is involved, it makes sense to follow [https://gist.github.com/mtrimarchi/b3b4addeb21081be6a561d8d90e051f5](url) as it solves the problem via `netplan`.
